File: mobile/android/app/src/main/java/com/volttracker/obdpoc/data/ObdQueryStore.java
package com.volttracker.obdpoc.data;

import com.volttracker.obdpoc.materialize.LocationSample;
import com.volttracker.obdpoc.materialize.MaterializerData;
import com.volttracker.obdpoc.materialize.PidObservation;
import com.volttracker.obdpoc.materialize.TelemetrySample;
import java.io.File;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Read-side of the on-device SQLite store. New callers that only need to read (typed records or
 * JSON projections) should depend on this narrower interface rather than the full {@link
 * ObdLocalStore} facade.
 *
 * <p>Both this and {@link ObdSessionStore} are implemented by {@link ObdLocalStore}; existing
 * callers can keep depending on the concrete facade. The split is additive; nothing was moved out
 * of {@link ObdLocalStore}.
 *
 * <p>Extends {@link MaterializerData} because the three {@code read*Samples} methods the
 * materializers consume are pure reads and belong on the query side.
 */
public interface ObdQueryStore extends MaterializerData {

    // ---- typed-record reads --------------------------------------------------------

    ObdSessionRecord getSession(long sessionId);

    List<ObdSessionRecord> getRecentSessions(int limit);

    List<TelemetrySampleRecord> getRecentTelemetry(long sessionId, int limit);

    List<StatusEventRecord> getRecentEvents(long sessionId, int limit);

    List<AdapterHistoryRecord> getAdapterHistory(int limit);

    // ---- JSON projections ----------------------------------------------------------

    StorageSummaryRecord getStorageSummaryRecord();

    JSONArray getRecentSessionsJson(int limit);

    JSONArray getAdapterHistoryJson(int limit);

    JSONArray getTripsJson(int limit);

    JSONObject getInsightsJson();

    // ---- materializer reads (from MaterializerData) --------------------------------

    @Override
    List<LocationSample> readLocationSamples(long sessionId);

    @Override
    List<PidObservation> readPidObservations(long sessionId);

    @Override
    List<TelemetrySample> readTelemetrySamples(long sessionId);

    // ---- DB file location (read-only metadata) -------------------------------------

    File getDatabaseFile();
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/data/ObdQueryStore.kt
package com.volttracker.obdpoc.data

import com.volttracker.obdpoc.materialize.LocationSample
import com.volttracker.obdpoc.materialize.MaterializerData
import com.volttracker.obdpoc.materialize.PidObservation
import com.volttracker.obdpoc.materialize.TelemetrySample
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Read-side of the on-device SQLite store. Callers that only need reads should depend on this
 * narrower interface rather than the full [ObdLocalStore] facade.
 */
interface ObdQueryStore : MaterializerData {
    fun getSession(sessionId: Long): ObdSessionRecord?

    fun getRecentSessions(limit: Int): List<ObdSessionRecord>

    fun getRecentTelemetry(
        sessionId: Long,
        limit: Int,
    ): List<TelemetrySampleRecord>

    fun getRecentEvents(
        sessionId: Long,
        limit: Int,
    ): List<StatusEventRecord>

    fun getAdapterHistory(limit: Int): List<AdapterHistoryRecord>

    fun getStorageSummaryRecord(): StorageSummaryRecord

    fun getStorageOverviewRecord(): StorageSummaryRecord

    fun getStorageDetailsJson(): JSONObject

    fun getRecentSessionsJson(limit: Int): JSONArray

    fun getAdapterHistoryJson(limit: Int): JSONArray

    fun getTripsJson(limit: Int): JSONArray

    fun getInsightsJson(): JSONObject

    override fun readLocationSamples(sessionId: Long): List<LocationSample>

    override fun readPidObservations(sessionId: Long): List<PidObservation>

    override fun readTelemetrySamples(sessionId: Long): List<TelemetrySample>

    fun getDatabaseFile(): File
}
